java/org/contikios/cooja/Simulation.java

```diff
@@ -150,68 +155,133 @@ public Simulation(Cooja cooja, long seed) {
-    lastStartRealTime = System.currentTimeMillis();
-    lastStartSimulationTime = getSimulationTimeMillis();
-    speedLimitLastRealtime = lastStartRealTime;
-    speedLimitLastSimtime = lastStartSimulationTime;
+    this.simulationThread = new Thread(() -> {
+      boolean isSimulationRunning = false;
+      boolean isAlive = true;
+      do {
+        EventQueue.Pair nextEvent = null;
+        try {
+          while (isAlive) {
+            Runnable cmd;
+            do {
+              cmd = isSimulationRunning ? commandQueue.poll() : commandQueue.take();
+              if (cmd == null) {
+                // No command to execute
+              } else if (cmd == START) {
+                if (!isSimulationRunning) {
+                  isSimulationRunning = true;
+                  setRunning(true);
+                }
+              } else if (cmd == STOP) {
+                if (isSimulationRunning) {
+                  isSimulationRunning = false;
+                  setRunning(false);
+                }
+              } else if (cmd == QUIT) {
+                isAlive = false;
+                isShutdown = true;
+                if (isSimulationRunning) {
+                  isSimulationRunning = false;
+                  setRunning(false);
+                }
+                // Process no more commands
+                break;
+              } else {
+                cmd.run();
+              }
+            } while (cmd != null);
+
+            if (isSimulationRunning) {
+              /* Handle one simulation event, and update simulation time */
+              nextEvent = eventQueue.popFirst();
+              assert nextEvent != null : "Ran out of events in eventQueue";
+              assert nextEvent.time >= currentSimulationTime : "Event from the past";
+              currentSimulationTime = nextEvent.time;
+              nextEvent.event.execute(currentSimulationTime);
+            }
+          }
+        } catch (RuntimeException e) {
+          if ("MSPSim requested simulation stop".equals(e.getMessage())) {
+            /* XXX Should be*/
+            logger.info("Simulation stopped due to MSPSim breakpoint");
+          } else {
+            logger.fatal("Simulation stopped due to error: " + e.getMessage(), e);
+            if (!Cooja.isVisualized()) {
+              /* Quit simulator if in test mode */
+              System.exit(1);
+            } else {
+              String title = "Simulation error";
+              if (nextEvent != null && nextEvent.event instanceof MoteTimeEvent moteTimeEvent) {
+                title += ": " + moteTimeEvent.getMote();
+              }
+              Cooja.showErrorDialog(title, e, false);
+            }
+          }
+        } catch (InterruptedException e) {
+          // Simulation thread interrupted - quit
+          logger.warn("simulation thread interrupted");
+          isAlive = false;
+          isShutdown = true;
+        } finally {
```

Comment:
Would removing the `finally` keyword and keeping the content of the block make any difference for this code?

Reply:
It should not. I will remove `finally`.